src/transformers/models/mplugdocowl/modeling_mplugdocowl.py

```diff
@@ -0,0 +1,599 @@
+            kernel_size=self.conv_shape,
+            stride=self.conv_shape,
+            bias=True,
+        )
+        ## align visual features with language embedding with fc
+        self.visual_fc = torch.nn.Linear(self.config.hreducer_hidden_size, config.text_config.hidden_size)
+        self.vit_eos = torch.nn.Parameter(torch.randn(1, 1, config.text_config.hidden_size))
+        self.gradient_checkpointing = False
+        self.post_init()
+
+    def forward(self, encoder_hidden_states=None):
+        r"""
+        Processes the encoder hidden states to reduce visual feature length and align them with language embeddings.
+
+        encoder_hidden_states  (`torch.FloatTensor` of shape `(batch_size, sequence_length, hidden_size)`, `optional`):
+            batch_size is the number of all images (global+crop) in a batch
+            Sequence of hidden-states at the output of the last layer of the encoder.
+
+            Returns:
+        torch.FloatTensor: The processed sequence output with reduced visual feature length and aligned with language embeddings.
+
+        """
+
+        encoder_hidden_states = encoder_hidden_states[:, 1:, :]  # remove the first cls token
+        B, L, C = encoder_hidden_states.shape  # B, 1024=(448/14)^2, 1024
+        H = int(torch.sqrt(torch.tensor(L)))
+        encoder_hidden_states = encoder_hidden_states.transpose(2, 1)
+
+        encoder_hidden_states = encoder_hidden_states.view(B, C, H, H)  # (BCHH)
+
+        hidden_states = self.reducer_before(encoder_hidden_states)  # B 4D H W/4
+        # hidden_states = self.reducer_activation(hidden_states)
+        B, XD, H, W_div_X = hidden_states.shape
+        X = self.conv_patch
+        D = XD // X
+
+        hidden_states = hidden_states.view(B, X, D, H, W_div_X)
+
+        hidden_states = hidden_states.permute(0, 2, 3, 4, 1)
+
+        hidden_states = hidden_states.reshape(B, D, H, W_div_X * X)
+
+        sequence_output = self.reducer(hidden_states)  # B,C,H,W -> B,C,H/conv_shape[0],W/(conv_shape[1])
+        sequence_output = sequence_output.flatten(2).transpose(
+            1, 2
+        )  # B,C,H/conv_shape[0],W/(conv_shape[1]) -> B,C,L/conv_patch -> B,L/conv_patch,C
+        sequence_output = sequence_output.transpose(0, 1).contiguous()  # L/conv_patch, B, C
+
+        sequence_output = self.visual_fc(sequence_output)  # L/conv_patch, B, h
+        sequence_output = sequence_output.transpose(0, 1).contiguous()  # B, s/4, h
+        sequence_output = torch.cat([sequence_output, self.vit_eos.repeat(B, 1, 1)], dim=1)
+
+        return sequence_output
+
+
+@add_start_docstrings(
+    """The MPLUGDOCOWL model which consists of a vision backbone and a language model.""",
+    MPLUGDOCOWL_START_DOCSTRING,
+)
+class MPLUGDocOwlForConditionalGeneration(MPLUGDocOwlPreTrainedModel):
+    def __init__(self, config: MPLUGDocOwlConfig):
+        super().__init__(config)
+
+        self.vision_tower = MPLUGDocOwlVisionModel(config.vision_config)
```

Comment:
Regarding OwlVit, Can you, please, specify on how we can recover q,k,v? To me it still seems different architecture. I suggest leaving as it is and opening new PR. Because I am not sure how much time it will take.